umap/utils.py

```diff
@@ -118,33 +118,30 @@ def is_ajax(request):
     return request.headers.get("x-requested-with") == "XMLHttpRequest"
 
 
-def merge_features(reference: list, latest: list, entrant: list):
-    # Just in case (eg. both removed the same element, or changed only metadatas)
-    if latest == entrant:
+class ConflictError(ValueError):
+    pass
+
+
+def merge_features(reference: list, latest: list, incoming: list):
+    """Finds the changes between reference and incoming, and reapplies them on top of latest."""
+    if latest == incoming:
         return latest
 
-    # Remove common features between entrant and reference versions (unchanged ones).
-    for feature in reference[:]:
-        for other in entrant[:]:
-            if feature == other:
-                entrant.remove(feature)
-                reference.remove(feature)
-                break  # no need to continue looking
-
-    # Now make sure remaining features are still in latest version
-    for feature in reference:
-        for other in latest:
-            if other == feature:
-                break
-        else:
-            # We cannot distinguish the case where both deleted the same
-            # element and added others, or where both modified the same
-            # element, so let's raise a conflict by caution.
-            raise ValueError
-
-    # We can merge.
-    for feature in reference:
-        latest.remove(feature)
-    for feature in entrant:
-        latest.append(feature)
-    return latest
+    reference_removed = [item for item in reference if item not in incoming]
```

Comment:
removed_from_reference ?

Reply:
Or just `removed` ? And then `added` ? Seems clear enough to me.

Reply:
Might be simpler yes 🙌